src/assets/dpos.ts

```diff
@@ -0,0 +1,137 @@
+ */
+
+import {
+	getLisk32AddressFromAddress,
+	getLisk32AddressFromPublicKey,
+} from '@liskhq/lisk-cryptography';
+import {
+	MODULE_NAME_DPOS,
+	INVALID_BLS_KEY,
+	DUMMY_PROOF_OF_POSSESSION,
+	INVALID_ED25519_KEY,
+	DPOS_INIT_ROUNDS,
+	// ROUND_LENGTH,
+} from '../constants';
+
+import {
+	AccountEntry,
+	BlockEntry,
+	ModuleResponse,
+	ValidatorEntry,
+	VoterEntry,
+	GenesisDataEntry,
+} from '../types';
+
+export const getValidatorKeys = async (blocks: BlockEntry[]): Promise<Record<string, string>> => {
+	const keys: Record<string, string> = {};
+	for (const block of blocks) {
+		const lskAddress: string = getLisk32AddressFromPublicKey(block.header.generatorPublicKey);
+		keys[lskAddress] = block.header.generatorPublicKey.toString('hex');
+		for (const trx of block.payload) {
+			const trxAddress: string = getLisk32AddressFromPublicKey(trx.senderPublicKey);
+			keys[trxAddress] = trx.senderPublicKey.toString('hex');
+		}
+	}
+	return keys;
+};
+
+export const createValidatorsArray = async (
+	accounts: AccountEntry[],
+	blocks: BlockEntry[],
+): Promise<ValidatorEntry[]> => {
+	const validators: ValidatorEntry[] = [];
+	const validatorKeys: any = await getValidatorKeys(blocks);
+
+	for (const account of accounts) {
+		if (account.dpos.delegate.username !== '') {
+			const validator: ValidatorEntry = {
+				address: getLisk32AddressFromAddress(account.address),
+				name: account.dpos.delegate.username,
+				blsKey: INVALID_BLS_KEY,
+				proofOfPossession: DUMMY_PROOF_OF_POSSESSION,
+				generatorKey: '',
+				lastGeneratedHeight: 0,
+				isBanned: false,
+				pomHeights: [],
+				consecutiveMissedBlocks: 0,
+			};
+
+			if (validatorKeys[account.address.toString('hex')]) {
+				validator.generatorKey = validatorKeys[account.address.toString('hex')];
+			} else {
+				validator.generatorKey = INVALID_ED25519_KEY;
+			}
+			validator.lastGeneratedHeight = account.dpos.delegate.lastForgedHeight;
+			validator.isBanned = account.dpos.delegate.isBanned;
+			validator.pomHeights = account.dpos.delegate.pomHeights;
+			validator.consecutiveMissedBlocks = account.dpos.delegate.consecutiveMissedBlocks;
+			validators.push(validator);
+		}
+	}
+	return validators;
+};
+
+export const createVotersArray = async (accounts: AccountEntry[]): Promise<VoterEntry[]> => {
+	const voters: VoterEntry[] = [];
+	for (const account of accounts) {
+		if (account.dpos.sentVotes && account.dpos.unlocking) {
+			const voter: VoterEntry = {
+				address: getLisk32AddressFromAddress(account.address),
+				sentVotes: account.dpos.sentVotes.map(vote => ({
+					delegateAddress: getLisk32AddressFromAddress(vote.delegateAddress),
+					amount: String(vote.amount),
+				})),
+				pendingUnlocks: account.dpos.unlocking.map(unlock => ({
+					delegateAddress: getLisk32AddressFromAddress(unlock.delegateAddress),
+					amount: String(unlock.amount),
+					unvoteHeight: unlock.unvoteHeight,
+				})),
+			};
+			voters.push(voter);
+		}
+	}
+	return voters;
```

Comment:
Do we need to sort `voters` as well?

Reply:
After a discussion on Slack we came to several conclusions:

1. It is important to ensure deterministic behavior for the genesis block.
2. The current version of the LIP already sort of specifies the order, under the condition that `for every` iterate by the array index (as I understand, this is not always guaranteed in JS). The reason being is that [this line](https://github.com/LiskHQ/lips/blob/77d1aa057d4f05300f3f7afaa96d8a43ceff086d/proposals/lip-0063.md?plain=1#L351) specifies the order for `accounts` and `legacyAccounts`, and [this line](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0030.md#dposobject-schema) guarantees that `account.pos.stakes` (`account.dpos.sentVotes`) is sorted.
3. It would be nice to explicitly sort the resulting arrays to minimize the possibility of an error and to be more explicit. In practice, the following arrays should be sorted by the `address` property of its elements (sometimes this property is nested): `legacyObj.accounts`, `authDataSubstore`, `validators`, `stakers` (=`voters`).

Reply:
Thank you for the confirmation, @sergeyshemyakov. 🙏 

@nagdahimanshu In conclusion, whenever we encounter a list involving `legacyObj.accounts`, `authDataSubstore`, `validators`, `stakers` let us explicitly sort them by the `address` property to ensure deterministic outputs. Please sort the `voters` list here as well 🙇

Reply:
Done!